ios-base/Navigators/Navigator.swift

```diff
@@ -0,0 +1,204 @@
+    case .modal:
+      currentViewController?.present(viewController, animated: animated, completion: completion)
+      currentViewController = viewController
+    case .push:
+      rootViewController?.pushViewController(viewController, animated: animated)
+      currentViewController = viewController
+    case .reset:
+      rootViewController?.setViewControllers([viewController], animated: animated)
+      currentViewController = viewController
+    case .changeRoot:
+      UIView.animate(withDuration: 0.3) { [weak self] in
+        if let navigation = (viewController as? UINavigationController) {
+          UIApplication.shared.keyWindow?.rootViewController = navigation
+          self?.rootViewController = navigation
+          self?.currentViewController = self?.rootViewController?.topViewController
+        } else {
+          let navigationController = UINavigationController(rootViewController: viewController)
+          UIApplication.shared.keyWindow?.rootViewController = navigationController
+          self?.rootViewController = navigationController
+          self?.rootViewController?.isNavigationBarHidden = true
+          self?.currentViewController = viewController
+        }
+      }
+    }
+  }
+
+  func navigate(to router: Navigator, animated: Bool, completion: (() -> Void)?) {
+    guard let viewController = router.rootViewController else {
+      assert(false, "Navigator does not have a root view controller")
+      return
+    }
+
+    currentViewController?.present(viewController, animated: animated, completion: completion)
+    currentViewController = viewController
+  }
+
+  func pop(animated: Bool = true) {
+    rootViewController?.popViewController(animated: animated)
+    currentViewController = rootViewController?.visibleViewController
+  }
+
+  func popToRoot(animated: Bool = true) {
+    rootViewController?.popToRootViewController(animated: animated)
+  }
+
+  func popTo(index: Int, animated: Bool = true) {
+    guard
+      let viewControllers = rootViewController?.viewControllers,
+      viewControllers.count > index
+      else { return }
+    let viewController = viewControllers[index]
+    rootViewController?.popToViewController(viewController, animated: animated)
+    currentViewController = viewController
+  }
+
+  func popTo(route: Route, animated: Bool = true) {
+    guard
+      let viewControllers = rootViewController?.viewControllers,
+      let viewController = viewControllers.first(where: {
+        type(of: $0) == type(of: route.screen)
+      })
+      else { return }
+    rootViewController?.popToViewController(viewController, animated: true)
+    currentViewController = viewController
+  }
+
+  func dismiss(animated: Bool = true, completion: (() -> Void)? = nil) {
+    let presentingViewController = currentViewController?.presentingViewController
+    currentViewController?.dismiss(animated: animated, completion: completion)
+    currentViewController = presentingViewController
```

Comment:
it seems like if the users directly calls dismiss, push, present the Navigator would loose track of the currentViewController. Maybe we can play with the idea of blocking this APIs, not sure if possible 🤔

Reply:
Yeah, that's a problem, i'm not sure if there's a way of listening to navigation events globally (with the notificationCenter for example)

Reply:
Ended up refactoring the currentViewController variable to be a computed one, since the rootViewController already tracks that from the 'visibleViewController` and `topViewController`